source/core/DualTypeVisitor.cpp

```diff
@@ -29,97 +29,99 @@ namespace Vireo
-                break;
-            case kEncoding_Enum:
-                success = EnumCompatible(typeRefX, typeRefY, operation);
-                break;
-            case kEncoding_Array: {
-                if (typeRefX->Rank() == 1 && typeRefX->GetSubElement(0)->BitEncoding() == kEncoding_Unicode) {
-                    success = StringCompatible(typeRefX, typeRefY);
-                } else {
-                    success = typeRefY->BitEncoding() == kEncoding_Array && ArrayCompatible(typeRefX, pDataX, typeRefY, pDataY, operation);
-                }
-                break;
-            }
-            case kEncoding_None:
-                success = typeRefY->BitEncoding() == kEncoding_None;  // Encountered when visiting an empty variant constants wired through SetAttribute
-                break;
-            default:
-                success = false;
+
+        EncodingEnum encodingX = typeRefX->BitEncoding();
+        switch (encodingX) {
+        case kEncoding_Boolean:
+            success = operation.AreBooleanCompatible(typeRefX, typeRefY);
+            break;
+        case kEncoding_UInt:
+            success = operation.AreUIntCompatible(typeRefX, typeRefY);
+            break;
+        case kEncoding_S2CInt:
+            success = operation.AreS2CIntCompatible(typeRefX, typeRefY);
+            break;
+        case kEncoding_IEEE754Binary:
+            success = operation.AreIEEE754BinaryCompatible(typeRefX, typeRefY);
+            break;
+        case kEncoding_Cluster:
+            success = typeRefY->BitEncoding() == kEncoding_Cluster && ClusterCompatible(typeRefX, pDataX, typeRefY, pDataY, operation);
+            break;
+        case kEncoding_Enum:
+            success = EnumCompatible(typeRefX, typeRefY, operation);
+            break;
+        case kEncoding_Array: {
+            if (typeRefX->Rank() == 1 && typeRefX->GetSubElement(0)->BitEncoding() == kEncoding_Unicode) {
+                success = StringCompatible(typeRefX, typeRefY);
+            } else {
+                success = typeRefY->BitEncoding() == kEncoding_Array && ArrayCompatible(typeRefX, pDataX, typeRefY, pDataY, operation);
             }
+            break;
+        }
+        case kEncoding_Variant:
+            success = VariantCompatible(typeRefX, pDataX, typeRefY, pDataY, operation);
+            break;
+        default:
+            success = false;
         }
         return success;
     }
 
     //------------------------------------------------------------
     bool DualTypeVisitor::VariantCompatible(TypeRef typeRefX, void* pDataX, TypeRef typeRefY, void* pDataY, const DualTypeOperation &operation)
     {
-        VariantTypeRef variantTypeX = reinterpret_cast<VariantTypeRef>(typeRefX);
-        VariantTypeRef variantTypeY = reinterpret_cast<VariantTypeRef>(typeRefY);
+        VariantDataRef variantX = *reinterpret_cast<VariantDataRef *>(pDataX);
```

Comment:
@siddhukrs , This is one notable functional change to keep this interface consistent in handling pointer to data. The caller should pass pointer to VariantDataRef.

Reply:
Yup, I think this is how I wanted it to be, but for some reason I was forced to use the TypeRef to access data.  This makes me more comfortable. Thanks for this change!